spec/index.bs

```diff
@@ -2424,114 +2424,110 @@ A {{GPUBindGroupLayout}} object has the following internal slots:
+                |descriptor|:
+            </pre>
 
-        <dfn>dynamic storage buffer validation</dfn>: There must be {{GPULimits/maxDynamicStorageBuffersPerPipelineLayout|GPULimits.maxDynamicStorageBuffersPerPipelineLayout}} or
-                fewer |bindingDescriptor|s of type {{GPUBindingType/storage-buffer}} with {{GPUBindGroupLayoutEntry/hasDynamicOffset}} set to `true`
-                in |descriptor| that are visible to any shader stage.
+            **Returns:** {{GPUBindGroupLayout}}
 
-        <dfn>sampled texture validation</dfn>: There must be {{GPULimits/maxSampledTexturesPerShaderStage|GPULimits.maxSampledTexturesPerShaderStage}} or
-            fewer |bindingDescriptor|s of type {{GPUBindingType/sampled-texture}} visible on each shader stage in |descriptor|.
+            1. Let |layout| be a new valid {{GPUBindGroupLayout}} object.
+            1. Issue the following steps on the [=Device timeline=] of |this|:
+                <div class=device-timeline>
+                    1. If any of the following conditions are unsatisfied:
+                        <div class=validusage>
+                            - |this| is a [=valid=] {{GPUDevice}}.
+                            - The {{GPUBindGroupLayoutEntry/binding}} of each entry in |descriptor| is unique.
+                            - For each shader stage, the number of entries in |descriptor| with
+                                {{GPUBindGroupLayoutEntry/type}} {{GPUBindingType/uniform-buffer}} &le;
+                                {{GPULimits/maxUniformBuffersPerShaderStage|GPULimits.maxUniformBuffersPerShaderStage}}.
+                            - For each shader stage, the number of entries in |descriptor| with
+                                {{GPUBindGroupLayoutEntry/type}} {{GPUBindingType/storage-buffer}} &le;
+                                {{GPULimits/maxStorageBuffersPerShaderStage|GPULimits.maxStorageBuffersPerShaderStage}}.
+                            - For each shader stage, the number of entries in |descriptor| with
+                                {{GPUBindGroupLayoutEntry/type}} {{GPUBindingType/sampled-texture}} &le;
+                                {{GPULimits/maxSampledTexturesPerShaderStage|GPULimits.maxSampledTexturesPerShaderStage}}.
+                            - For each shader stage, the number of entries in |descriptor| with
+                                {{GPUBindGroupLayoutEntry/type}} {{GPUBindingType/readonly-storage-texture}} or
+                                {{GPUBindingType/writeonly-storage-texture}} &le;
+                                {{GPULimits/maxStorageTexturesPerShaderStage|GPULimits.maxStorageTexturesPerShaderStage}}.
+                            - For each shader stage, the number of entries in |descriptor| with
+                                {{GPUBindGroupLayoutEntry/type}} {{GPUBindingType/sampler}} &le;
+                                {{GPULimits/maxSamplersPerShaderStage|GPULimits.maxSamplersPerShaderStage}}.
```

Comment:
Lot of duplicated language here, maybe we can reduce it. Either by un-inlining the per-shader-stage limits and the per-pipeline-layout limits and referring to them here, or maybe just by tweaking the language. Or even by writing it as a table?